platform/entanglement/src/services/CopyService.js

```diff
@@ -54,45 +57,156 @@ define(
+            });
+        };
 
-            if (!this.validate(domainObject, parent)) {
+        /**
+         * Will persist a list of {@link objectClones}. It will persist all
+         * simultaneously, irrespective of order in the list. This may
+         * result in automatic request batching by the browser.
+         * @private
+         * @param progress
+         * @returns {Function} a function that will perform the persistence
+         * with a progress callback curried into it.
+         */
+        CopyService.prototype.persistObjects = function(progress) {
+            var persisted = 0,
+                self = this;
+            return function(objectClones) {
+                return self.$q.all(objectClones.map(function(clone, index){
+                    clone.model.persisted = self.now();
+                    return self.persistenceService.createObject(clone.persistenceSpace, clone.id, clone.model)
+                        .then(function(){
+                            return progress && progress({phase: "copying", totalObjects: objectClones.length, processed: ++persisted});
+                        });
+                })).then(function(){
+                    return objectClones;
+                });
+            };
+        };
+
+        /**
+         * Will add a list of clones to the specified parent's composition
+         * @private
+         * @param parent
+         * @param progress
+         * @returns {Function}
+         */
+        CopyService.prototype.addClonesToParent = function(parent, progress) {
+            var self = this;
+            return function(clones) {
+                var parentClone = clones[clones.length-1];
+                if (!parent.hasCapability('composition')){
+                    return self.$q.reject();
+                }
+
+                return self.persistenceService
+                    .updateObject(parentClone.persistenceSpace, parentClone.id, parentClone.model)
+                    .then(function(){return parent.getCapability("composition").add(parentClone.id);})
+                    .then(function(){return parent.getCapability("persistence").persist();})
+                    .then(function(){return parentClone;});
+                    // Ensure the clone of the original domainObject is returned
+            };
+        };
+
+        /**
+         * Creates a duplicate of the object tree starting at domainObject to
+         * the new parent specified.
+         * @param domainObject
+         * @param parent
+         * @param progress
+         * @returns a promise that will be completed with the clone of
+         * domainObject when the duplication is successful.
+         */
+        CopyService.prototype.perform = function (domainObject, parent) {
+            var $q = this.$q,
+                deferred = $q.defer();
+            if (this.validate(domainObject, parent)) {
+                this.buildCopyPlan(domainObject, parent, deferred.notify)
```

Comment:
I'd softly recommend creating a `CopyTask` class or similar to encapsulate this process. There are a few new private methods on `CopyService` that all need to pass around a similar set of objects as arguments, which can be a little hard to follow and may be error-prone; think you could move this all into a `new CopyTask(domainObject, parent, deferred).run()`. Internally, `CopyTask` would have `buildCopyPlan`, `persistObjects`, `addClonesToParent`, but they'd all be looking at instance variables instead of arguments, making it a little easier to follow/maintain (instead of looking at each method and its usages to figure out what's what, just go to the constructor)